### ISSUE TITLE
Compiler doesn't seem to catch all dependencies

### ISSUE THREAD

**Issue by tdprogramming** [apache/royale-asjs#297]
Hi, I have a new project I'm developing using Royale called XFL Tools, it's designed to parse assets created with Adobe Animate "out of the box" and render them using a replicated Flash API:

https://github.com/tdprogramming/xfltools

I've listed the issues I'm having in the read me file. The main problem is that the compiler seems to leave out needed dependencies, probably due to trying to eliminate circular dependencies? When trying to run the sample project out-of-the-box, for example, the Shape class required by the Greensock Tween libs (used as a ticker) doesn't have its superclasses added as dependencies and so fails to init properly. Manually adding the full hierarchy as dependencies using the Python script I devised addresses this, but it's not a long term solution of course.

Is it possible to add hints for the compiler or force certain dependencies through?

The second issue is that I need to patch the generated HTML file to include the create JS framework and a few other things. Is it possible to inject HTML into the output, for example by using tagged comments in the source? I seem to remember reading about this back when Royale was Flex JS, but haven't been able to get it to work.

Many thanks for any insight you can give on these issues.

**Comment by piotrzarzycki21** [comment]
Hello @tdprogramming ,

Here is the example how you can inject some script. I'm injecting library for coloring code on the website [Highlight](https://github.com/piotrzarzycki21/TranspiledActionScript/blob/examples/Examples/TranspiledActionScriptWebsite/src/utils/Highlight.as)

**Comment by aharui** [comment]
Are you on the latest nightly build?  Some dependency issues were fixed on 9/14.

If you are still having problems, attach some files that show the source and output.  Or better yet, a reduced example.  An external superclass (where there is no AS for the superclass since it comes from the Greensock JS file) should not be in the requires list and will need to be loaded by your template before the Royale code starts to run.

**Comment by tdprogramming** [comment]
Thank you for your helpful responses.

I can't get the inject_html tags in the comments to work - do they need to be in a certain place (eg. before a constructor?). However, from your examples I saw how to use a custom HTML template. This allowed me to do the injection that way.

I'm using the 0.9.4 build from today (9/17/2018). A lot of the dependency issues have gone - there is just one, I'm using an EventDispatcher class to mimic the Actionscript EventDispatcher. It uses a Vector of Event listeners, and the compiler doesn't add the requirement for Language.js automatically. 

It's hard to simplify the sample code itself as the "Main" function is doing very little, but spins up my mimicked DisplayObject classes which inherit back to EventDispatcher. If you download the full checkout at the framework site: https://github.com/tdprogramming/xfltools

You can open the sample project in Moonshine IDE and compile for JS-debug (I need to remove my public vars for release build). If you run the JS-Debug build in Google Chrome, you'll see the console throws an error trying to spin up a Vector in EventDispatcher. You can run the patchIndex.py script on the output JS to fix this. 

Note that this sample needs to run in a browser or simulated server debug environment such as WAMP, in order to load the XFL file.

I'll keep investigating and try to get some simpler sample code.

**Comment by aharui** [comment]
Yes, inject_html has to be on the constructor.  At least for now.  That way we don't have to move it from the class to the constructor.  The class tokens don't end up in the output JS but the constructor does.

I will look into the Vector problem.  You might be able to work around it by adding some other code to that class that also forces Language.js.

**Comment by aharui** [comment]
In my simple test of Vector (which used String instead of EventListener), the output did reference Language.js.  Make sure you deleted any .JS files in the output folders.  Royale does not copy over existing JS files from SWCs, only for AS files it is transpiling, so if you change a file in a SWC, the application's copy of that file may be stale.

Otherwise, post the output you are getting.

**Comment by aharui** [comment]
I tried static too and got the right thing.  Again, make sure you are not ending up with a stale copy and then attach that file.  Static Dependencies should be listed as such in the output.

**Comment by tdprogramming** [comment]
I've attached a zip of the project as compiled clean. You can run the output here:

tdprogramming.com/demos/xfltools_staging

First error reported by Google Chrome console is:

Uncaught TypeError: Cannot read property 'trace' of undefined
    at flash.display.Shape.flash.events.EventDispatcher [as constructor] (EventDispatcher.js:27)
    at Function.childCtor.base (base.js:2514)
    at flash.display.Shape.flash.display.DisplayObject [as constructor] (DisplayObject.js:24)
    at Function.childCtor.base (base.js:2514)
    at flash.display.Shape.flash.display.InteractiveObject [as constructor] (InteractiveObject.js:22)
    at Function.childCtor.base (base.js:2514)
    at flash.display.Shape.flash.display.DisplayObjectContainer [as constructor] (DisplayObjectContainer.js:22)
    at Function.childCtor.base (base.js:2514)
    at flash.display.Shape.flash.display.Sprite [as constructor] (Sprite.js:22)
    at Function.childCtor.base (base.js:2514)

(note that this sample won't run from a local disk unless you have your permissions set to allow it to load external files).

[js-debug.zip](https://github.com/apache/royale-asjs/files/2397912/js-debug.zip)

**Comment by aharui** [comment]
OK, I was able to run it locally and see the problem.  It looks like a static initializer can make it's base class's non-static dependencies required at static initialization time.  Might take a while to figure out how to resolve it.

**Comment by tdprogramming** [comment]
Many thanks for looking into it!

**Comment by aharui** [comment]
Here is the minimal test case.  A static dependency or base class has its own static dependency or base class that has non-static dependencies in the constructor.  The constructor is called during evaluation of static dependencies which is before non-static dependencies have been loaded.


[TestCase.zip](https://github.com/apache/royale-asjs/files/2399265/TestCase.zip)

**Comment by Harbs** [comment]
FWIW, the simplest way to resolve these dependencies is to delay static instantiations until all classes are loaded.

So, we add a new file loaded after all others (i.e. `static_inst.js`) with the following for the test case:
````
/**
 * @private
 */
Main._myStaticClass = new StaticClass();

/**
 * @export
 * @type {AnotherStaticClass}
 */
StaticClass.another = new AnotherStaticClass();
````
Done like this, all statics would be be removed from the main class file (and possibly just declared) and order should not matter.

The only case that I can think of which might cause issues is dependencies on static instantiations within constructors.

**Comment by aharui** [comment]
I'm not convinced that you can always defer static initialization.  I believe an instance of a base class is created as the prototype of a class and that base class instance runs its constructor very early.  If that constructor needs to access the static var, I don't think the initialization file will have run before we set up class prototypes.

**Comment by Harbs** [comment]
I'm pretty sure I manually pulled out the instantiations into a separate file when I had this problem in the past to see if this would work and it did.

I'm not following you. base() is only run inside the constructor. Where do you think the constructors will be run early? Unless you have a case where you have circular static (or semi-static) refs in constructors, I'm having trouble coming up with scenarios where this will not work. But I could be missing something...

**Comment by aharui** [comment]
Let's add BaseClass.as to the scenario with StaticClass extends BaseClass.  And BaseClass looks something like:

public class BaseClass
{
   private static var counter:Counter = new Counter();

    public function BaseClass()
    {
        Counter.increment("BaseClass"); // count the number of instances.
    }
}

I thought somewhere (like goog.inherits) the prototype is set up.  In plain JS, the code for StaticClass would do something like:

StaticClass = function() {};
StaticClass.prototype = new BaseClass();
StaticClass.prototype.someFunction = function

I think the "new BaseClass()" call will fail because Counter has not been setup.

**Comment by Harbs** [comment]
I don't think so.

Here's goog.inherits:

````

goog.inherits = function(childCtor, parentCtor) {
  /** @constructor */
  function tempCtor() {}
  tempCtor.prototype = parentCtor.prototype;
  childCtor.superClass_ = parentCtor.prototype;
  childCtor.prototype = new tempCtor();
  /** @override */
  childCtor.prototype.constructor = childCtor;

  /**
   * Calls superclass constructor/method.
   *
   * This function is only available if you use goog.inherits to
   * express inheritance relationships between classes.
   *
   * NOTE: This is a replacement for goog.base and for superClass_
   * property defined in childCtor.
   *
   * @param {!Object} me Should always be "this".
   * @param {string} methodName The method name to call. Calling
   *     superclass constructor can be done with the special string
   *     'constructor'.
   * @param {...*} var_args The arguments to pass to superclass
   *     method/constructor.
   * @return {*} The return value of the superclass method/constructor.
   */
  childCtor.base = function(me, methodName, var_args) {
    // Copying using loop to avoid deop due to passing arguments object to
    // function. This is faster in many JS engines as of late 2014.
    var args = new Array(arguments.length - 2);
    for (var i = 2; i < arguments.length; i++) {
      args[i - 2] = arguments[i];
    }
    return parentCtor.prototype[methodName].apply(me, args);
  };
};
````
It does some fancy footwork to *not* call the base constructor.

**Comment by aharui** [comment]
Cool!  Good to know.  So I agree that should work although I'm still wondering about how to determine the order of snippets.   How did you manage these static snippets?  I'm trying to figure out if it is worth creating a separate file in the SWC or whether we need to scan all JS files in the "Publisher".  But meanwhile, I think I have a less invasive change that might work that I am going to try.
